File: backend/api/exercises.py
"""
Exercise Timer API Endpoints
============================

FastAPI endpoints for the Exercise Timer feature
(exercises.exercise_timers / exercises.exercise_attempts).

Endpoints (contract from .features/designs_active/007-001_design.md):
    GET    /api/exercises                    -> { data: ExerciseTimerSummary[], count }
    GET    /api/exercises/{exercise_id}      -> { exercise, last_attempt }
    POST   /api/exercises                    -> created ExerciseTimer (201); 409 duplicate name
    PUT    /api/exercises/{exercise_id}      -> updated ExerciseTimer; 404 / 409
    DELETE /api/exercises/{exercise_id}      -> 200; cascades the timer's attempts (OQ-1)
    POST   /api/exercises/{exercise_id}/attempts -> saved ExerciseAttempt (201)
"""

from datetime import datetime

from fastapi import APIRouter, HTTPException

from backend_functions.queries import (
    get_exercise_timer,
    get_last_attempt,
    list_exercise_summaries,
    create_exercise,
    update_exercise,
    delete_exercise,
    create_attempt,
    ExerciseError,
    ExerciseNotFoundError,
    ExerciseNameConflictError,
    ExerciseValidationError,
)
from backend.schemas.exercises_schemas import (
    ExerciseAttemptCreateRequest,
    ExerciseCreateRequest,
    ExerciseUpdateRequest,
)

router = APIRouter(prefix="/api/exercises", tags=["exercises"])


# ---------------------------------------------------------------------------
# Serialization
# ---------------------------------------------------------------------------

_TIMER_DATETIME_FIELDS = ("created_at", "updated_at")
_ATTEMPT_DATETIME_FIELDS = ("started_at", "ended_at", "created_at")


def _serialize_timer(timer: dict) -> dict:
    """Normalize timer NUMERIC interval + datetime fields for JSON."""
    if timer is None:
        return None
    out = dict(timer)
    if out.get("interval_seconds") is not None:
        out["interval_seconds"] = float(out["interval_seconds"])
    for field in _TIMER_DATETIME_FIELDS:
        if isinstance(out.get(field), datetime):
            out[field] = out[field].isoformat()
    return out


def _serialize_attempt(attempt: dict) -> dict:
    """Normalize attempt NUMERIC interval + datetime fields for JSON."""
    if attempt is None:
        return None
    out = dict(attempt)
    if out.get("interval_seconds_used") is not None:
        out["interval_seconds_used"] = float(out["interval_seconds_used"])
    for field in _ATTEMPT_DATETIME_FIELDS:
        if isinstance(out.get(field), datetime):
            out[field] = out[field].isoformat()
    return out


def _serialize_summary(summary: dict) -> dict:
    """Normalize a summary row: timer fields + derived aggregate numerics."""
    out = dict(summary)
    if out.get("interval_seconds") is not None:
        out["interval_seconds"] = float(out["interval_seconds"])
    for field in _TIMER_DATETIME_FIELDS:
        if isinstance(out.get(field), datetime):
            out[field] = out[field].isoformat()
    # MAX() over an empty set yields NULL; keep as-is so the UI sees null stats.
    return out
# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.get("")
async def list_timers():
    """List every timer with per-timer attempt aggregates (OQ-2), ordered by name."""
    try:
        data = [_serialize_summary(s) for s in list_exercise_summaries()]
        return {"data": data, "count": len(data)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch exercise timers: {str(e)}")


@router.get("/{exercise_id}")
async def get_timer(exercise_id: int):
    """Get a single timer with its most recent attempt (progress-bar calibration)."""
    try:
        timer = get_exercise_timer(exercise_id)
        if timer is None:
            raise HTTPException(status_code=404, detail="Exercise timer not found")
        last_attempt = None
        if timer is not None:
            last_attempt = get_last_attempt(exercise_id)
        return {
            "exercise": _serialize_timer(timer),
            "last_attempt": _serialize_attempt(last_attempt),
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch exercise timer: {str(e)}")


@router.post("", status_code=201)
async def create_timer(req: ExerciseCreateRequest):
    """Create a new exercise timer. 409 on case-insensitive duplicate name."""
    try:
        timer = create_exercise(
            name=req.name,
            interval_seconds=req.interval_seconds,
            notes=req.notes,
        )
        return _serialize_timer(timer)
    except ExerciseNameConflictError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except ExerciseError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create exercise timer: {str(e)}")


@router.put("/{exercise_id}")
async def update_timer(exercise_id: int, req: ExerciseUpdateRequest):
    """Edit an existing timer (partial update). 404 if missing, 409 on duplicate name."""
    try:
        timer = update_exercise(
            exercise_id=exercise_id,
            name=req.name,
            interval_seconds=req.interval_seconds,
            notes=req.notes,
        )
        return _serialize_timer(timer)
    except ExerciseNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ExerciseNameConflictError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except ExerciseError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update exercise timer: {str(e)}")


@router.delete("/{exercise_id}")
async def delete_timer(exercise_id: int):
    """
    Permanently delete a timer and ALL of its attempt history (OQ-1).
    Always returns 200 on success; the cascade runs in one transaction (T02).
    """
    try:
        delete_exercise(exercise_id)
        return {"success": True}
    except ExerciseNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ExerciseError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete exercise timer: {str(e)}")


@router.post("/{exercise_id}/attempts", status_code=201)
async def save_attempt(exercise_id: int, req: ExerciseAttemptCreateRequest):
    """
    Save one confirmed attempt. Only user-confirmed paced_count/total_count and
    the interval snapshot are written (FR-8, FR-10).
    """
    try:
        attempt = create_attempt(
            exercise_id=exercise_id,
            started_at=req.started_at,
            ended_at=req.ended_at,
            interval_seconds_used=req.interval_seconds_used,
            paced_count=req.paced_count,
            total_count=req.total_count,
            notes=req.notes,
        )
        return _serialize_attempt(attempt)
    except ExerciseNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ExerciseValidationError as e:
        raise HTTPException(status_code=422, detail=str(e))
    except ExerciseError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to save exercise attempt: {str(e)}")

File: frontend/pifitness/src/lib/types/exercises.ts
/**
 * Exercise Timer Types
 *
 * Cross-surface data contract for the Exercise Timer feature.
 * Shared between frontend (React) and backend (FastAPI Pydantic models).
 *
 * Maps directly to the pre-built backend tables (no migrations; design 007-001):
 *   exercises.exercise_timers    (master data — one row per named exercise)
 *   exercises.exercise_attempts  (one row per saved Start/Stop cycle)
 *
 * Endpoints (backend/api/exercises.py, prefix /api/exercises):
 *   GET    /api/exercises               -> { data: ExerciseTimerSummary[], count }
 *   GET    /api/exercises/{id}          -> { exercise, last_attempt }
 *   POST   /api/exercises               -> created ExerciseTimer; 409 on duplicate name
 *   PUT    /api/exercises/{id}          -> updated ExerciseTimer; 409 on duplicate name
 *   DELETE /api/exercises/{id}          -> cascades the timer and ALL of its attempts
 *                                          (attempts deleted first, then the timer row —
 *                                          OQ-1 app-layer transaction honoring ON DELETE RESTRICT)
 *   POST   /api/exercises/{id}/attempts -> saved ExerciseAttempt
 *
 * A run-in-progress lives entirely in the client (no "in-progress" row); only
 * confirmed values from the save prompt are persisted.
 */

/**
 * A named exercise timer (master data).
 * Maps to exercises.exercise_timers.
 */
export interface ExerciseTimer {
  /** Primary key (SERIAL). */
  exercise_id: number;
  /** Display name, unique case-insensitively (idx_exercise_timers_name). */
  name: string;
  /**
   * Pacing interval in seconds. NUMERIC(5,2) CHECK > 0.
   * UI input restricts to one decimal of precision; the DB accepts up to two.
   */
  interval_seconds: number;
  /** Optional notes (unused in v1). */
  notes?: string | null;
  /** Row creation time. */
  created_at: string;
  /** Last update time (server now() on UPDATE). */
  updated_at: string;
}

/**
 * A saved attempt (one Start/Stop cycle).
 * Maps to exercises.exercise_attempts.
 */
export interface ExerciseAttempt {
  /** Primary key (SERIAL). */
  attempt_id: number;
  /** Parent timer id (ON DELETE RESTRICT — app deletes attempts first on timer delete). */
  exercise_id: number;
  /**
   * Snapshot of the timer's interval in effect during this attempt,
   * NUMERIC(5,2) NOT NULL. Editing the timer later must not re-pace history.
   */
  interval_seconds_used: number;
  /** ISO timestamp when the run started (after the 5→1 countdown). */
  started_at: string;
  /** ISO timestamp when Stop was pressed. */
  ended_at: string;
  /** Reps completed while keeping pace. CHECK >= 0. */
  paced_count: number;
  /** Reps completed overall. CHECK >= paced_count. */
  total_count: number;
  /** Optional notes (unused in v1). */
  notes?: string | null;
  /** Row creation time. */
  created_at: string;
}

/**
 * A timer enriched with per-timer attempt aggregates for the selection screen.
 * Derived fields (not columns):
 *   - last_attempt_total_count: total_count of the most recent attempt
 *     (per `(exercise_id, started_at DESC)` index lookup)
 *   - last_attempt_paced_count: paced_count of that same attempt
 *   - highest_score: MAX(total_count) across attempts (OQ-2 — the "highest-ever" score)
 *   - highest_paced_count: MAX(paced_count) across attempts (OQ-2 — both efforts tracked)
 */
export interface ExerciseTimerSummary extends ExerciseTimer {
  /** Total reps of the most recent attempt, if any. */
  last_attempt_total_count: number | null;
  /** On-pace reps of the most recent attempt, if any. */
  last_attempt_paced_count: number | null;
  /** All-time-highest total reps (OQ-2). */
  highest_score: number | null;
  /** All-time-highest on-pace reps (OQ-2). */
  highest_paced_count: number | null;
}

/** Response for GET /api/exercises (timer summaries with attempt aggregates). */
export interface ExerciseListResponse {
  data: ExerciseTimerSummary[];
  count: number;
}

/** Response for GET /api/exercises/{id}. */
export interface ExerciseDetailResponse {
  /** The timer record. */
  exercise: ExerciseTimer;
  /** The most recent attempt (used to calibrate the live progress bar), or null. */
  last_attempt: ExerciseAttempt | null;
}

/** Request body for POST /api/exercises (create a timer). */
export interface ExerciseCreateRequest {
  /** Display name (unique case-insensitively). */
  name: string;
  /** Pacing interval in seconds, > 0, one-decimal UI precision. */
  interval_seconds: number;
  /** Optional notes (unused in v1). */
  notes?: string | null;
}

/** Request body for PUT /api/exercises/{id} (edit a timer). */
export interface ExerciseUpdateRequest {
  /** Display name (unique case-insensitively). */
  name?: string;
  /** Pacing interval in seconds, > 0, one-decimal UI precision. */
  interval_seconds?: number;
  /** Optional notes (unused in v1). */
  notes?: string | null;
}

/**
 * Request body for POST /api/exercises/{id}/attempts (save a confirmed attempt).
 * The client-side countdown/count-up defaults are NEVER sent here — only the
 * user-confirmed paced_count and total_count, and the interval snapshot.
 */
export interface ExerciseAttemptCreateRequest {
  /** ISO timestamp when the run started (after the 5→1 countdown). */
  started_at: string;
  /** ISO timestamp when Stop was pressed. */
  ended_at: string;
  /** Snapshot of the timer's interval in effect during the run. */
  interval_seconds_used: number;
  /** User-confirmed on-pace reps. CHECK >= 0. */
  paced_count: number;
  /** User-confirmed total reps. CHECK >= paced_count. */
  total_count: number;
  /** Optional notes (unused in v1). */
  notes?: string | null;
}